rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2160 @@
+
+#### Requirements on Owned Values
+[transmute-owned]: #transmuting-owned-values
+
+Transmutations involving owned values must adhere to two rules to be sound. They must:
+ * [preserve or broaden the bit validity][owned-validity], and
+ * [preserve or shrink the size][owned-size].
+
+##### Preserve or Broaden Bit Validity
+[owned-validity]: #Preserve-or-Broaden-Bit-Validity
+[`NonZeroU8`]: https://doc.rust-lang.org/beta/core/num/struct.NonZeroU8.html
+
+The bits of any valid instance of the source type must be a bit-valid instance of the destination type.
+
+For example, we are permitted to transmute a `Bool` into a [`u8`]:
+```rust
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(u8)]
+enum Bool {
+    True = 1,
+    False = 0,
+}
+
+let _ : u8 = Bool::True.transmute_into();
+let _ : u8 = Bool::False.transmute_into();
+```
+<sup>
+(Note: <code>#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]</code> annotation connotes that <i>all</i> aspects of <code>Bool</code>'s layout are part of its <a href="#When-is-a-transmutation-stable?">library stability guarantee</a>.)
+</sup>
+
+...because all possible instances of `Bool` are also valid instances of [`u8`]. However, transmuting a [`u8`] into a `Bool` is forbidden:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _ : Bool = u8::default().transmute_into(); // Compile Error!
+```
+...because not all instances of [`u8`] are valid instances of `Bool`.
+
+Another example: While laying out certain types, Rust may insert padding bytes between the layouts of fields. In the below example `Padded` has two padding bytes, while `Packed` has none:
+```rust
+#[repr(C)]
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+struct Padded(pub u8, pub u16, pub u8);
+
+#[repr(C)]
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+struct Packed(pub u16, pub u16, pub u16);
+
+assert_eq!(mem::size_of::<Packed>(), mem::size_of::<Padded>());
+```
+
+We may safely transmute from `Packed` to `Padded`:
+```rust
+let _ : Padded = Packed::default().transmute_into();
+```
+...but not from `Padded` to `Packed`:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _ : Packed = Padded::default().transmute_into(); // Compile Error!
+```
+...because doing so would expose two uninitialized padding bytes in `Padded` as if they were initialized bytes in `Packed`.
+
+##### Preserve or Shrink Size
+[owned-size]: #Preserve-or-Shrink-Size
+
+It's completely sound to transmute into a type with fewer bytes than the source type; e.g.:
+```rust
+let _ : [u8; 16] = [0u8; 32]::default().transmute_into();
+```
+This transmute truncates away the final sixteen bytes of the `[u8; 32]` value.
+
+A value may ***not*** be transmuted into a type of greater size:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _ : [u8; 32] = [0u8; 16]::default().transmute_into(); // Compile Error!
+```
+
```

Comment:
This may be obvious, but this doesn't necessarily make it clear *why* it can't be transmuted into a type of greater size. The previous example makes it clear that doing the transmute would expose two uninitialized padding bytes. I presume the same problem is true in this case (the extra bytes would be uninitialized) but it may be worth calling that out for clarity.

Reply:
> I presume the same problem is true in this case (the extra bytes would be uninitialized) but it may be worth calling that out for clarity.

This is an interesting observation, because I hadn't considered that expanding the size could *ever* be valid, except your observation suggests a case where it could be: transmuting `[MaybeUninit<u8>; 2]` to `[MaybeUninit<u8>; 4]` (or any situation where the trailing bit of the destination type may be uninit)!

I'm going to rephrase this as:
> A value may not be transmuted into a type of greater size, **if doing so would expose uninitialized bytes as if they were initialized**:

Reply:
Resolved by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/1a119b21bc3ca7c447c1bd645cb0805d0c2b5b3c